%pip install --force-reinstall --no-cache-dir numpy pandas matplotlib seaborn

In [2]:
import pandas as pd 
import numpy as np
import matplotlib as plt
import seaborn as sns

import warnings 
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv('D:\Customer-Churn-Analytics\Data\Superstore.csv', encoding='latin1')
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

### Data Quality
* Found mismatched Datatype in date columns
* Found 0 Nulls and Duplicates
* we have 9994 rows and 21 columns 

In [5]:
df.isnull().sum()
df.duplicated().sum()

np.int64(0)

In [13]:
df.shape

(9994, 21)

### Categorical and numerical columns 
* Found 15 categorical columns and 6 numerical columns



In [ ]:
df.select_dtypes('object').columns 

Index(['order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_id',
       'customer_name', 'segment', 'country', 'city', 'state', 'region',
       'product_id', 'category', 'sub_category', 'product_name'],
      dtype='str')

In [ ]:
df.select_dtypes('number').columns

Index(['row_id', 'postal_code', 'sales', 'quantity', 'discount', 'profit'], dtype='str')

###  Renaming the column names adding removing spaces and adding '_'

In [ ]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
df.columns = df.columns.str.lower().str.replace('-', '_')
df.columns

Index(['row_id', 'order_id', 'order_date', 'ship_date', 'ship_mode',
       'customer_id', 'customer_name', 'segment', 'country', 'city', 'state',
       'postal_code', 'region', 'product_id', 'category', 'sub_category',
       'product_name', 'sales', 'quantity', 'discount', 'profit'],
      dtype='str')

### Change Datatypes


In [15]:
date_columns = ['order_date', 'ship_date',]

for col in date_columns:
    df[col] = pd.to_datetime(df[col])


In [16]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   row_id         9994 non-null   int64         
 1   order_id       9994 non-null   str           
 2   order_date     9994 non-null   datetime64[us]
 3   ship_date      9994 non-null   datetime64[us]
 4   ship_mode      9994 non-null   str           
 5   customer_id    9994 non-null   str           
 6   customer_name  9994 non-null   str           
 7   segment        9994 non-null   str           
 8   country        9994 non-null   str           
 9   city           9994 non-null   str           
 10  state          9994 non-null   str           
 11  postal_code    9994 non-null   int64         
 12  region         9994 non-null   str           
 13  product_id     9994 non-null   str           
 14  category       9994 non-null   str           
 15  sub_category   9994 non-null   s

### Blank/Whitespace Strings in Object Columns

In [17]:
text_cols = df.select_dtypes(include=['object']).columns
for col in text_cols:
    empty_strings = df[col].astype(str).str.strip().eq('').sum()
    if empty_strings > 0:
        print(f"Column '{col}' has {empty_strings} empty/whitespace string entries.")
        

In [18]:
df.loc[df['sales']<0] 
df.loc[df['quantity']<0] 
df.loc[df['discount']<0] 
df.loc[df['profit']<0].head(5) # found negative numbers in profit 

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,postal_code,region,product_id,category,sub_category,product_name,sales,quantity,discount,profit
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
14,15,US-2015-118983,2015-11-22,2015-11-26,Standard Class,HP-14815,Harold Pawlan,Home Office,United States,Fort Worth,...,76106,Central,OFF-AP-10002311,Office Supplies,Appliances,Holmes Replacement Filter for HEPA Air Cleaner...,68.8100,5,0.80,-123.8580
15,16,US-2015-118983,2015-11-22,2015-11-26,Standard Class,HP-14815,Harold Pawlan,Home Office,United States,Fort Worth,...,76106,Central,OFF-BI-10000756,Office Supplies,Binders,Storex DuraTech Recycled Plastic Frosted Binders,2.5440,3,0.80,-3.8160
23,24,US-2017-156909,2017-07-16,2017-07-18,Second Class,SF-20065,Sandra Flanagan,Consumer,United States,Philadelphia,...,19140,East,FUR-CH-10002774,Furniture,Chairs,"Global Deluxe Stacking Chair, Gray",71.3720,2,0.30,-1.0196
27,28,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,...,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royal...",3083.4300,7,0.50,-1665.0522


### Saving the cleaned file

In [27]:
cols = ['row_id','order_id', 'order_date', 'ship_date', 'ship_mode', 'customer_id',
       'customer_name', 'segment', 'country', 'city', 'state', 'postal_code',
       'region', 'product_id', 'category', 'sub_category', 'product_name',
       'sales', 'quantity', 'discount', 'profit', 'cost', 'profit_margin',
       'shipping_days', 'order_year', 'order_month', 'order_quarter',
       'order_days']

for col in cols:
    max_char_len = df[col].astype(str).str.len().max()
    print(f"{col}: {max_char_len}")

row_id: 4
order_id: 14
order_date: 10
ship_date: 10
ship_mode: 14
customer_id: 8
customer_name: 22
segment: 11
country: 13
city: 17
state: 20
postal_code: 5
region: 7
product_id: 15
category: 15
sub_category: 11
product_name: 127
sales: 9
quantity: 2
discount: 4
profit: 10
cost: 18
profit_margin: 19
shipping_days: 1
order_year: 4
order_month: 2
order_quarter: 1
order_days: 2


In [19]:
df.to_csv('D:\Customer-Churn-Analytics\Data\cleaned_superstore.csv', index=False)

### Feature Engineering 

In [20]:
# cost column
df['cost'] = (df['sales'] - df['profit'])



In [21]:
# profit_margin column
df['profit_margin'] = (df['profit'] / df['sales'] * 100)

In [22]:
df['shipping_days'] = (df['ship_date'] - df['order_date']).dt.days


In [23]:
df['order_year'] = df['order_date'].dt.year
df['order_month'] = df['order_date'].dt.month
df['order_quarter'] = df['order_date'].dt.quarter
df['order_days'] = df['order_date'].dt.day

In [24]:
df

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,quantity,discount,profit,cost,profit_margin,shipping_days,order_year,order_month,order_quarter,order_days
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,2,0.00,41.9136,220.0464,16.00,3,2016,11,4,8
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,3,0.00,219.5820,512.3580,30.00,3,2016,11,4,8
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,2,0.00,6.8714,7.7486,47.00,4,2016,6,2,12
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,5,0.45,-383.0310,1340.6085,-40.00,7,2015,10,4,11
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,2,0.20,2.5164,19.8516,11.25,7,2015,10,4,11
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9989,9990,CA-2014-110422,2014-01-21,2014-01-23,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,3,0.20,4.1028,21.1452,16.25,2,2014,1,1,21
9990,9991,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,2,0.00,15.6332,76.3268,17.00,5,2017,2,1,26
9991,9992,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,2,0.20,19.3932,239.1828,7.50,5,2017,2,1,26
9992,9993,CA-2017-121258,2017-02-26,2017-03-03,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,4,0.00,13.3200,16.2800,45.00,5,2017,2,1,26


In [25]:
df.to_csv('D:\Customer-Churn-Analytics\Data\cleaned_superstore.csv', index=False)
df.to_csv('D:\Customer-Churn-Analytics\Data\Featured_superstore.csv', index=False)